jupyterhub/auth.py

```diff
@@ -631,11 +666,51 @@ def _default_executor(self):
         """
      ).tag(config=True)
 
+    admin_groups = Set(
+        help="""
+        User groups that determine admin access. Users not in these groups can still
+        be granted admin status through admin_users. While/blacklisting rules still apply.
+        """
+    ).tag(config=True)
+
     def __init__(self, **kwargs):
         if pamela is None:
             raise _pamela_error from None
         super().__init__(**kwargs)
 
+    @run_on_executor
+    def is_admin(self, handler, data):
```

Comment:
I see the problem now.

- `@run_on_executor` wraps a *blocking* method and makes it return a Future
- calling `super().is_admin()` can only be called from a true coroutine

So the options are:

1. make the base `.is_admin()` method blocking so that it can be called here without `await`
2. split this method so that `.is_admin` is a true coroutine (`async def is_admin`), and it `awaits` super, then calls `._check_groups`, a private `@run_on_executor` method.

Either one's fine with me.

Reply:
Thank you so much, I would have never figured that out on my own. I opted for the former. The basic is_admin in Authenticator should be trivial enough that it doesn't need to be async'd.